test/rosdep_repo_check/alpine.py

```diff
@@ -0,0 +1,112 @@
+#       contributors may be used to endorse or promote products derived from
+#       this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import os
+import tarfile
+from io import TextIOWrapper
+
+from . import open_gz_url
+from . import PackageEntry
+from . import RepositoryCacheCollection
+
+
+def replace_tokens(string, os_code_name, os_arch):
+    """Replace tokens in the repository base URL."""
+    for key, value in {
+        '$basearch': os_arch,
+        '$releasever': os_code_name,
+    }.items():
+        string = string.replace(key, value)
+    return string
+
+
+def parse_apkindex(f):
+    # An example of an APKINDEX entry. Entries are divided by a blank line.
+
+    # V:2.1.1-r0
+    # A:x86_64
+    # S:6645
+    # I:28672
+    # T:RTP proxy (documentation)
+    # U:https://www.rtpproxy.org/
+    # L:BSD-2-CLause
+    # o:rtpproxy
+    # m:Natanael Copa <ncopa@alpinelinux.org>
+    # t:1602354892
+    # c:183e99f73bb1223768aa7231a836a3e98e94c03e
+    # i:docs rtpproxy=2.1.1-r0
+
+    fio = TextIOWrapper(f)
+    while True:
+        entry = {}
+        while (l := fio.readline()) != '\n':
+            k, v = l.strip().split(':', 1)
+            entry[k] = v
+
+        yield entry
+
+
+def enumerate_alpine_packages(base_url, os_name, os_code_name, os_arch):
+    """
+    Enumerate packages in an Alpine repository.
+
+    :param base_url: the Alpine repository base URL.
+    :param os_name: the name of the OS associated with the repository.
+    :param os_code_name: the OS version associated with the repository.
+    :param os_arch: the system architecture associated with the repository.
+
+    :returns: an enumeration of package entries.
+    """
+
+    base_url = replace_tokens(base_url, os_code_name, os_arch)
+    # get main or community or testing from the second last component of the url
+    pkg_source = os.path.split(os.path.split(base_url)[0])[1]
+    apkindex_url = os.path.join(base_url, 'APKINDEX.tar.gz')
+    print('Reading Alpine package metadata from ' + apkindex_url)
+
+    with open_gz_url(apkindex_url) as f:
+        with tarfile.open(fileobj=f) as tf:
+            index = tf.extractfile("APKINDEX")
```

Comment:
This doesn't seem to work correctly to parse the stream of data coming down and tries to seek around like a file on disk. After some manipulation, I was able to get it working with this change:
```suggestion
        with tarfile.open(mode='r|', fileobj=f) as tf:
            index = None
            for ti in tf:
                if ti.name == 'APKINDEX':
                    index = tf.extractfile(ti)
                    break
            if index is None:
                raise RuntimeError('APKINDEX url did not contain an APKINDEX file')
```